src/input/shared.rs

```diff
@@ -136,3 +137,22 @@ pub fn float_as_int<'a>(input: &'a impl Input<'a>, float: f64) -> ValResult<'a,
         Err(ValError::new(ErrorTypeDefaults::IntParsingSize, input))
     }
 }
+
+pub fn decimal_as_int<'a>(py: Python, input: &'a impl Input<'a>, decimal: &PyAny) -> ValResult<'a, EitherInt<'a>> {
+    if !decimal.call_method0(intern!(py, "is_finite"))?.extract::<bool>()?
+        || decimal.call_method0(intern!(py, "is_nan"))?.extract::<bool>()?
+    {
+        return Err(ValError::new(ErrorTypeDefaults::FiniteNumber, input));
+    }
+    let (numerator, denominator) = decimal
+        .call_method0(intern!(py, "as_integer_ratio"))?
+        .extract::<(BigInt, BigInt)>()?;
+    if denominator != 1.into() {
+        return Err(ValError::new(ErrorTypeDefaults::IntFromFloat, input));
+    }
+    if let Ok(i) = i64::try_from(&numerator) {
+        Ok(EitherInt::I64(i))
+    } else {
+        Ok(EitherInt::BigInt(numerator))
+    }
```

Comment:
Suggestion: skip trying to extract the numerator here and can avoid needing a `BigInt` for the denominator; if it's not `1` we don't need to successfully extract it.

```suggestion
    let (numerator, denominator) = decimal
        .call_method0(intern!(py, "as_integer_ratio"))?
        .extract::<(&PyAny, &PyAny)>()?;
    if denominator.extract::<i64>().map_or(true, |d| d != 1) {
        return Err(ValError::new(ErrorTypeDefaults::IntFromFloat, input));
    }
    Ok(EitherInt::Py(numerator))
```